### PR TITLE
BITAU-108 Add unlockWithAuthenticatorSyncKey to VaultRepository 

### PR DESCRIPTION
**Pull request by ahaisting-livefront**
## 🎟️ Tracking

https://livefront.atlassian.net/browse/BITAU-108

## 📔 Objective

I had previously missed this requirement for this ticket:

 "A function is added to `VaultRepository` / `VaultLockManager` that allows for this per-user key to be used to unlock a particular user's vault."

## 📸 Screenshots

N/A

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt [hunk 1]
```diff
@@ -553,6 +553,24 @@ class VaultRepositoryImpl(
...
+            .getAuthenticatorSyncUnlockKey(userId = userId)
+            ?: return VaultUnlockResult.InvalidStateError
+        return unlockVaultForUser(
+            userId = userId,
+            initUserCryptoMethod = InitUserCryptoMethod.DecryptedKey(
+                decryptedUserKey = authenticatorSyncKey,
+            ),
+        )
+            .also {
+                if (it is VaultUnlockResult.Success) {
```
Comment (dseverns-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt:568]: Cool, we are kind of using this scope function the same way you could use the `onSuccess(action)` for the Kotlin`Result`?! Nice 😎

**Review by dseverns-livefront — DISMISSED** [review]

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt (hunk 1 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt:568]: Totally! Can't say this was my idea, I was inspired by a few other functions, but I'm digging it.

**Review by SaintPatrck — DISMISSED** [review]

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt [hunk 2]
```diff
@@ -553,6 +553,24 @@ class VaultRepositoryImpl(
         mutableTotpCodeResultFlow.tryEmit(totpCodeResult)
     }
 
+    override suspend fun unlockVaultWithAuthenticatorSyncKey(): VaultUnlockResult {
+        val userId = activeUserId ?: return VaultUnlockResult.InvalidStateError
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt:557]: Is this going to work as something that is based on the current user? We'll need to unlock the vaults for all users that have enabled the feature. Also, I'm not sure it should necessarily be an error if the current user does not have a key for that same reason...it's a valid state to not have a key.

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt [hunk 3]
```diff
@@ -167,6 +167,18 @@ interface VaultRepository : CipherManager, VaultLockManager {
...
+     * Attempt to unlock the vault using the stored authenticator sync unlock key for the
+     * given user.
+     *
+     * @param userId ID of the user's vault to unlock.
+     * @param decryptedAuthenticatorSyncKey The authenticator sync unlock key for the user.
+     */
+    suspend fun unlockVaultWithAuthenticatorSyncKey(
+        userId: String,
+        decryptedAuthenticatorSyncKey: String,
+    ): VaultUnlockResult
```
Comment (ahaisting-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt:180]: Updated this to be per user and also require an unlock key at calltime! @brian-livefront

Comment (brian-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt:180]: Does the naming here need to be specific to the authenticator sync feature? Like, this could be any of the user master keys right? The POC uses:

```kotlin
    suspend fun unlockVaultWithDecryptedKey(
        userId: String,
        decryptedUserKey: String,
    ): VaultUnlockResult
```

I _think_ that is the correct terminology based on the SDK. I would go in that direction probably because I can see the potential for this call to be used for keys derived in other ways as well.

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt [hunk 4]
```diff
@@ -553,6 +553,23 @@ class VaultRepositoryImpl(
...
+    ): VaultUnlockResult {
+        return unlockVaultForUser(
+            userId = userId,
+            initUserCryptoMethod = InitUserCryptoMethod.DecryptedKey(
+                decryptedUserKey = decryptedAuthenticatorSyncKey,
+            ),
+        )
+            .also {
+                if (it is VaultUnlockResult.Success) {
+                    deriveTemporaryPinProtectedUserKeyIfNecessary(userId = userId)
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt:568]: Just curious...why is this necessary here?

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt (hunk 2 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt]: We spoke about this a bit offline, too. 

TLDR; we want this to be per user, and the branch is updated as such.

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt (hunk 3 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt]: Indeed, the SDK uses `decryptedUserKey`! 

My intention here was to try to lead hints about where a caller of this function is supposed to be getting the key from. But I'm happy to update to the SDK's terminology: 055c8f3f6f24e37d2d6a1452f0554d11a11809b9

The param doc leaves a hint anyway:

```
@param decryptedUserKey The authenticator sync unlock key for the user.
```

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt (hunk 4 above)
Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryImpl.kt]: 👍 Thanks!

Yeah when I was looking this over earlier I clearly didn't fully grok what was going on. From my reading, it seemed like it would be possible that this could have an impact but, looking it at it now, `deriveTemporaryPinProtectedUserKeyIfNecessary` has nothing to do with what we are doing. Admittedly, the `ifNecessary` suffix on the function name made me ✨feel✨  safe 😬 


I removed the `.also` call and updated test accordingly: 4230b7363313ec7502f94c7fdb7c51bd1836ac08

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt [hunk 5]
```diff
@@ -167,6 +167,18 @@ interface VaultRepository : CipherManager, VaultLockManager {
      */
     fun emitTotpCodeResult(totpCodeResult: TotpCodeResult)
 
+    /**
+     * Attempt to unlock the vault using the stored authenticator sync unlock key for the
+     * given user.
+     *
+     * @param userId ID of the user's vault to unlock.
+     * @param decryptedUserKey The authenticator sync unlock key for the user.
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt:175]: Can we maybe combine the general and specific approaches and do something like

```kotlin
     * @param decryptedUserKey A decrypted user key for the user (ex: their authenticator sync unlock key)
```

Comment (ahaisting-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt]: Yeah, I totally see what you mean here...

What do you think of this:

780faf2d099ef9b9e4e2cd43922691cc77c57fa7

Keeps things generic to match the vault sdk and also to allow for future functionality, but leaves hints about what this is currently used for.

Comment (brian-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt]: Looks good!

File: app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt [hunk 6]
```diff
@@ -167,6 +167,18 @@ interface VaultRepository : CipherManager, VaultLockManager {
...
     fun emitTotpCodeResult(totpCodeResult: TotpCodeResult)
 
+    /**
+     * Attempt to unlock the vault using a user unlock key.
+     *
+     * @param userId ID of the user's vault to unlock.
+     * @param decryptedUserKey A decrypted unlock key for the user (ex: their authenticator
+     * sync unlock key)
+     */
+    suspend fun unlockVaultWithDecryptedUserKey(
```
Comment (brian-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/vault/repository/VaultRepository.kt:177]: 👍

File: app/src/test/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryTest.kt [hunk 7]
```diff
@@ -1243,6 +1243,100 @@ class VaultRepositoryTest {
             }
         }
 
+    @Test
+    @Suppress("MaxLineLength")
+    fun `unlockVaultWithAuthenticatorSyncKey with VaultLockManager Success should return Success`() =
```
Comment (brian-livefront) [inline app/src/test/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryTest.kt:1248]: You need to update the test names now.

Comment (ahaisting-livefront) [reply app/src/test/java/com/x8bit/bitwarden/data/vault/repository/VaultRepositoryTest.kt:1248]: 🤦 thank you. 

71357ee2f2fc39c91a6b85a3d798cd7b50f21b85

**Review by brian-livefront — APPROVED** [review]

**Comment by ahaisting-livefront** [comment]
Thanks for the comments, @brian-livefront !
